examples/src/main/java/kafka/examples/ExactlyOnceMessageProcessor.java

```diff
@@ -0,0 +1,190 @@
+            }
+
+            consumer.assign(topicPartitions);
+            printWithTxnId("Manually assign partitions: " + topicPartitions);
+        }
+
+        int messageProcessed = 0;
+        while (messageRemaining.get() > 0) {
+            ConsumerRecords<Integer, String> records = consumer.poll(Duration.ofMillis(200));
+            if (records.count() > 0) {
+                try {
+                    // Begin a new transaction session.
+                    producer.beginTransaction();
+                    for (ConsumerRecord<Integer, String> record : records) {
+                        // Process the record and send to downstream.
+                        ProducerRecord<Integer, String> customizedRecord = transform(record);
+                        producer.send(customizedRecord);
+                    }
+                    Map<TopicPartition, OffsetAndMetadata> positions = new HashMap<>();
+                    for (TopicPartition topicPartition : consumer.assignment()) {
+                        positions.put(topicPartition, new OffsetAndMetadata(consumer.position(topicPartition), null));
+                    }
+                    // Checkpoint the progress by sending offsets to group coordinator broker.
+                    // Under group mode, we must apply consumer group metadata for proper fencing.
+                    if (this.mode.equals("groupMode")) {
+                        producer.sendOffsetsToTransaction(positions, consumer.groupMetadata());
+                    } else {
+                        producer.sendOffsetsToTransaction(positions, consumerGroupId);
+                    }
+
+                    // Finish the transaction. All sent records should be visible for consumption now.
+                    producer.commitTransaction();
+                    messageProcessed += records.count();
+                } catch (CommitFailedException e) {
+                    // In case of a retriable exception, suggest aborting the ongoing transaction first for correctness.
+                    producer.abortTransaction();
+                } catch (ProducerFencedException | FencedInstanceIdException e) {
+                    throw new KafkaException("Encountered fatal error during processing: " + e.getMessage());
+                }
+            }
+            messageRemaining.set(messagesRemaining(consumer));
+            printWithTxnId("Message remaining: " + messageRemaining);
+        }
+
+        printWithTxnId("Finished processing " + messageProcessed + " records");
+        latch.countDown();
+    }
+
+    private void printWithTxnId(final String message) {
+        System.out.println(transactionalId + ": " + message);
+    }
+
+    private ProducerRecord<Integer, String> transform(final ConsumerRecord<Integer, String> record) {
+        printWithTxnId("Transformed record (" + record.key() + "," + record.value() + ")");
+        return new ProducerRecord<>(outputTopic, record.key() / 2, "Transformed_" + record.value());
```

Comment:
Why divide the key by two?

Reply:
It's just a way of showing the key gets processed by message copier, all the produced message key are even keys.